### PR TITLE
Avoid shelling out multiple times when activating Ruby

### PR DESCRIPTION
**Pull request by rzane**
We can load all of the environment information we'll need in a single invocation of the ruby executable.

### CODE REVIEW

**Comment by vinistock** [comment]
While I agree with the idea to avoid invoking the Ruby executable excessively, this breaks up a bit for shadowenv, since it has a different flow than the other version managers. The shadowenv extension updates the process environment, but we never update the Ruby version and YJIT attributes.

Maybe we can unify those two only?

File: src/ruby.ts [hunk 1]
```diff
@@ -36,24 +36,29 @@ export class Ruby {
...
-      .getExtension("shopify.vscode-shadowenv")
-      ?.activate();
+  private async activateFallback() {
+    const shadowenv = vscode.extensions.getExtension(
+      "shopify.vscode-shadowenv"
+    );
+
+    if (shadowenv) {
+      await shadowenv.activate();
+      await this.delay(500);
```
Comment (rzane) [inline src/ruby.ts:58]: This delay should only be necessary for shadowenv.

File: src/ruby.ts [hunk 2]
```diff
@@ -36,24 +36,29 @@ export class Ruby {
...
         `Failed to activate ${this.versionManager} environment: ${error.message}`
       );
     }
   }
 
-  private async activateShadowenv() {
-    await vscode.extensions
-      .getExtension("shopify.vscode-shadowenv")
-      ?.activate();
+  private async activateFallback() {
```
Comment (vinistock) [inline src/ruby.ts:51]: Let's continue to call it `activateShadowenv`.

File: src/ruby.ts [hunk 3]
```diff
@@ -80,19 +85,28 @@ export class Ruby {
...
-    const rubyVersion = await asyncExec('ruby -e "puts RUBY_VERSION"');
-    this.rubyVersion = rubyVersion.stdout.trim();
-    vscode.window.setStatusBarMessage(`Ruby ${this.rubyVersion}`);
+  private async analyze(
+    ruby: string,
+    opts?: ExecOptions
+  ): Promise<{ [key: string]: string }> {
+    const result = await asyncExec(
+      `${ruby} -rjson -e "puts JSON.dump(env: ENV.to_h, version: RUBY_VERSION, yjit: defined?(RubyVM::YJIT))"`,
+      { ...opts, encoding: "utf-8" }
```
[the remark this answers is not in the document]
Comment (rzane) [reply src/ruby.ts:101]: I added this to appease typescript. If I don't specify the encoding, `result.stdout` is `string | Buffer`. If I specify utf-8, `result.stdout` is `string`.

File: src/ruby.ts (hunk 2 above)
Comment (vinistock) [reply src/ruby.ts:51]: Another option is to add another branch for shadowenv and have the default do nothing. In that case, we won't be using a version manager and will just pick up whatever is currently set in the VS Code environment.

**Comment by vinistock** [comment]
Thank you for the contribution, but I think main diverged considerably since then.

**Pull request closed without merging**